pkg/chunk/cache/fifo_cache.go

```diff
@@ -179,91 +191,130 @@ func (c *FifoCache) Put(ctx context.Context, keys []string, values []interface{}
 	}
 }
 
+func (c *FifoCache) addToHead(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	entry.prev, entry.next = nil, c.first
+	if c.first != nil {
+		c.first.prev = entry
+	}
+	c.first = entry
+	if c.last == nil {
+		c.last = entry
+	}
+}
+
+func (c *FifoCache) deleteFromTail() *cacheEntry {
+	if c.last == nil {
+		return nil
+	}
+	ret := c.last
+	c.last = ret.prev
+	if c.last == nil {
+		c.first = nil
+	} else {
+		c.last.next = nil
+	}
+	return ret
+}
+
+func (c *FifoCache) deleteFromList(entry *cacheEntry) {
+	if entry == nil {
+		return
+	}
+	if entry.prev == nil {
+		c.first = entry.next
+	} else {
+		entry.prev.next = entry.next
+	}
+	if entry.next == nil {
+		c.last = entry.prev
+	} else {
+		entry.next.prev = entry.prev
+	}
+}
+
 func (c *FifoCache) put(ctx context.Context, key string, value interface{}) {
 	// See if we already have the entry
-	index, ok := c.index[key]
+	entry, ok := c.entries[key]
 	if ok {
-		entry := c.entries[index]
-		deltaSize := sizeOf(value) - sizeOf(entry.value)
-
-		entry.updated = time.Now()
-		entry.value = value
-
 		// Remove this entry from the FIFO linked-list.
-		c.entries[entry.prev].next = entry.next
-		c.entries[entry.next].prev = entry.prev
-
-		// Corner case: updating last element
-		if c.last == index {
-			c.last = entry.prev
+		c.deleteFromList(entry)
+		entrySz := sizeOf(entry)
+		c.currSize -= entrySz
+
+		// New entry size
+		entrySz += (sizeOfInterface(value) - sizeOfInterface(entry.value))
+
+		if c.maxSizeBytes > 0 && c.currSize+entrySz > c.maxSizeBytes {
+			// New entry does not fit in the cache. Delete the entry.
```

Comment:
This is not how an LRU should work. We should evicted the last-recently used entries in the cache until we've freed enough space for this entry.

An easier way to implement this `put()` function is just keeping the "delete the existing item from list" in this `if ok` branch, and then just let re-add it to the data structure below, so that the eviction logic below is triggered also for the case the entry already exists.

Reply:
At this point I have already deleted the item from the LRU. This code handles a special (unlikely) case where we already have an item with this key in the cache, but the new item is too big to fit the cache.

Reply:
I may argue this is not an edge case. After some time you run the cache, the cache will run always close to the limit (because we start evicting only once the cache is full), and so hitting this case doesn't look unlikely to me.

Reply:
You are right, thank you for catching this!